lib/voters.js

```diff
@@ -39,188 +40,136 @@ module.exports = api;
+    try {
+      voter = await storage.voters.get({ledgerNodeId, privateKey, publicKey});
+    } catch(e) {
+      if(_createKeyErrorFilter(e)) {
+        // error indicates a retry may pass, throw it to try again
+        throw e;
       }
-      // simultaneous calls to this API can lead to multiple workers attempting
-      // to create keys causing _createVoter to produce a DuplicateError
-      function errorFilter(err) {
-        return ['DuplicateError', 'LockError'].includes(err.name);
+
+      if(e.name !== 'NotFoundError') {
+        // error is not recoverable, abort retrying
+        throw new pRetry.AbortError(e.message);
       }
-      async.retry(
-        {interval: 50, times: 100, errorFilter}, callback => storage.voters.get(
-          {ledgerNodeId, privateKey, publicKey}, (err, voter) => {
-            if(err && err.name === 'NotFoundError') {
-              const lockKey = `keygenerationlock|${ledgerNodeId}`;
-              return async.auto({
-                lock: callback => cache.client.setnx(
-                  lockKey, '', (err, result) => {
-                    if(err) {
-                      return callback(err);
-                    }
-                    if(result === 0) {
-                      return callback(new BedrockError(
-                        'Key generation in process. Try again.', 'LockError'));
-                    }
-                    // lock was sucssful
-                    callback();
-                  }),
-                create: ['lock', (results, callback) =>
-                  _createVoter(ledgerNodeId, callback)],
-                unlock: ['create', (results, callback) =>
-                  cache.client.del(lockKey, callback)]
-              }, (err, results) => {
-                if(err) {
-                  return callback(err);
-                }
-                callback(null, results.create);
-              });
-            }
-            callback(err, voter);
-          }), callback);
-    }],
-    cacheInsert: ['voter', (results, callback) => {
-      if(results.cache) {
-        return callback();
+
+      // voter not found, try to create it...
+
+      // lock to do key generation
+      const lockKey = `keygenerationlock|${ledgerNodeId}`;
+      const result = await cache.client.setnx(lockKey, '');
+      if(result === 0) {
+        // wait for about the length of time it takes to generate a key
+        // and then throw to try again
+        await _wait(KEY_GENERATION_TIME);
+        throw new BedrockError(
+          'Key generation in process. Try again.', 'LockError');
       }
-      const {id} = results.voter;
-      cache.client.set(voterKey, id, callback);
-    }]
-  }, (err, results) => {
-    if(err) {
-      return callback(err);
+
+      // lock acquired, create voter
+      voter = await _createVoter(ledgerNodeId);
+
+      // release lock
+      await cache.client.del(lockKey);
     }
-    callback(null, results.voter);
-  });
-};
+  }, {retries: 100});
```

Comment:
The behavior of these p-retry params is *far* different than the previous async.retry!  p-retry is *exponential* and starts at a 1s delay vs the previous constant 50ms interval retries.  The last attempt here will be closer to never than 5s. ;-)
https://github.com/tim-kos/node-retry#retryoperationoptions
https://caolan.github.io/async/docs.html#retry